debug jinja function only core / fusion CLI (#8448)
fishtown-analytics/docs.getdbt.com: website/docs/reference/dbt-jinja-functions/debug-method.md

---

> :::warning CLI only - Not available in dbt platform

Author: ```suggestion
:::warning Requires local development environment
```

> The `debug()` macro is only available when using <Constant name="core" /> CLI or <Constant name="fusion_engine" /> CLI in a local development environment. It is **not available** in <Constant name="platform" />. Do not deploy code to production that uses the `debug` macro.

Author: ```suggestion
The `debug()` macro is only available when using <Constant name="core" /> CLI or <Constant name="fusion_engine" /> CLI in a local development environment. It is **not available** in <Constant name="dbt_platform" />. Do not deploy code to production that uses the `debug` macro.
```

> The `debug()` macro is only available when using <Constant name="core" /> CLI or <Constant name="fusion_engine" /> CLI in a local development environment. It is **not available** in <Constant name="dbt_platform" />. Do not deploy code to production that uses the `debug` macro.

Reviewer: this is great! i tested this with fusion cli and it looks like it's partially supported in the sense of it exists and executes, but doesn't open an interactive python debugger and the output isn't executable code (useless jinja output basically). so it isn't useful. i don't think it should be fully documented but should we raise this as a fusion issue or are they aware that this doesn't work fully? the reason why i'm wondering we should mention this is because the current behavior could affect compile speed, creates massive files, and doesn't actually help or provide value?

testing below and found in compiled sql:

```
state {
    name: "jaffle_shop.test_debug_macro",
    current_block: None,
    auto_escape: None,
    ctx: {
        "kwargs": {},
        "some_number": 42,
        "test_variable": 'Hello from debug test',
        "varargs": (),
        "debug": minijinja::functions::builtins::debug, 
        "test_debug_macro": <macro test_debug_macro>,
        "MACRO_DISPATCH_ORDER": {},
        "TARGET_PACKAGE_NAME": 'jaffle_shop',
        "TARGET_UNIQUE_ID": 'model.jaffle_shop.test_debug_model',
        "__minijinja_current_execution_phase": 'render',
        "__minijinja_current_path": 'macros/test_debug.sql',
        "__minijinja_current_span"
```

Reviewer: comamnd used: `dbtf compile --select test_debug_model`

Reviewer: also confirmed in internal fusion slack: https://dbt-labs.slack.com/archives/C08PYQS8B7Z/p1768822157770869?thread_ts=1768822070.910159&cid=C08PYQS8B7Z

> The `debug()` macro is only available when using <Constant name="core" /> CLI or <Constant name="fusion_engine" /> CLI in a local development environment. It is **not available** in <Constant name="dbt_platform" />. Do not deploy code to production that uses the `debug` macro.

Reviewer: do we want to give platform users another option that works in a diff way? for eg:
```suggestion
The `debug()` macro is only available when using <Constant name="core" /> CLI or <Constant name="fusion_engine" /> CLI in a local development environment. It's _not available_ in <Constant name="dbt_platform" />. 

Do not deploy code to production that uses the `debug` macro.

If developing in <Constant name="dbt_platform"/>, you can also use:
- [`{{ print() }}`](/reference/dbt-jinja-functions/print) - Print messages to both the log file and standard output (`stdout`).
- [`{{ log() }}`](reference/dbt-jinja-functions/log) - Structured logging that prints messages during Jinja rendering.

```

Reviewer: sugg removing fusion cli in this sugg https://github.com/dbt-labs/docs.getdbt.com/pull/8448#discussion_r2704385060

> The `debug()` macro is only available when using <Constant name="core" /> CLI or <Constant name="fusion_engine" /> CLI in a local development environment. It is **not available** in <Constant name="cloud" /> (<Constant name="cloud_ide" />, <Constant name="cloud_cli" />, or <Constant name="orchestrator" />). Do not deploy code to production that uses the `debug` macro.

Author: ```suggestion
The `debug()` macro is only available when using <Constant name="core" /> CLI or <Constant name="fusion_engine" /> CLI in a local development environment. It is **not available** in <Constant name="platform" />. Do not deploy code to production that uses the `debug` macro.
```

:::

The `{{ debug() }}` macro will open an iPython debugger in the context of a compiled dbt macro. The `DBT_MACRO_DEBUGGING` environment variable must be set to use the debugger.

This function requires:

> - Interactive terminal access with iPython debugger (`ipdb`) installed

Reviewer: i think we should confirm if fusion supports a ipython debugger since fusion is built on rust and doesn't use ipdb. e.g.:

```suggestion
- Interactive terminal access with iPython debugger (`ipdb`) installed. <Constant name="fusion"/> doesn't provide a iPython (ipdb) debugger since its built on Rust. It instead outputs a non-interactive snapshot of the MiniJinja render context in the compiled code.
```

> - Local development environment running <Constant name="core" /> CLI or <Constant name="fusion_engine" /> CLI

Reviewer: removing if we think it's less confusing for the user to not use debug in fusion.
```suggestion
- Local development environment running <Constant name="core" /> CLI
```

---

> :::warning Requires local development environment

Author: ```suggestion
:::warning Requires Core CLI
```

> The `debug()` macro is only available when using <Constant name="core" /> CLI or <Constant name="fusion_engine" /> CLI in a local development environment. It's _not available_ in <Constant name="dbt_platform" />. 

Reviewer: i think we should remove fusion here?
```suggestion
The `debug()` macro is only available when using <Constant name="core" /> CLI in a local development environment. It's _not available_ in <Constant name="dbt_platform" /> or <Constant name="fusion"/>. 
```

> The `debug()` macro is only available when using <Constant name="core" /> CLI or <Constant name="fusion_engine" /> CLI in a local development environment. It's _not available_ in <Constant name="dbt_platform" />. 

Author: ```suggestion
The `debug()` macro is only available when using <Constant name="core" /> CLI in a local development environment. It's _not available_ in <Constant name="dbt_platform" />. 
```

Do not deploy code to production that uses the `debug` macro.

> If developing in <Constant name="dbt_platform"/>, you can also use:

Reviewer: ```suggestion
If developing in <Constant name="dbt_platform"/> or using <Constant name="fusion"/>, you can instead use:
```

- [`{{ print() }}`](/reference/dbt-jinja-functions/print) - Print messages to both the log file and standard output (`stdout`).

> - [`{{ log() }}`](reference/dbt-jinja-functions/log) - Structured logging that prints messages during Jinja rendering.

Reviewer: oops i forgot to forward slash
```suggestion
- [`{{ log() }}`](/reference/dbt-jinja-functions/log) - Structured logging that prints messages during Jinja rendering.
```
